{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "dc7defee-872c-4a35-82e0-7cebbc0cda56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/ntw/Desktop/boltz1_inputs/P21731_Tirzepatide.yaml\n",
      "version: 1\n",
      "sequences:\n",
      "  - protein:\n",
      "\tid: A\n",
      "\tsequence: YAEGTFTSDYSIALDKIAQKAFVQWLIAGGPSSGAPPPS\n",
      "\tmsa: --use_msa_server\n",
      "  - protein:\n",
      "\tid: B\n",
      "\tsequence: MWPNGSSLGPCFRPTNITLEERRLIASPWFAASFCVVGLASNLLALSVLAGARQGGSHTRSSFLTFLCGLVLTDFLGLLVTGTIVVSQHAALFEWHAVDPGCRLCRFMGVVMIFFGLSPLLLGAAMASERYLGITRPFSRPAVASQRRAWATVGLVWAAALALGLLPLLGVGRYTVQYPGSWCFLTLGAESGDVAFGLLFSMLGGLSVGLSFLLNTVSVATLCHVYHGQEAAQQRPRDSEVEMMAQLLGIMVVASVCWLPLLVFIAQTVLRNPPAMSPAGQLSRTTEKELLIYLRVATWNQILDPWVYILFRRAVLRRLQPRLSTRPRSLSLQPQLTQRSGLQ\n",
      "\tmsa: --use_msa_server\n",
      "\tmodifications:\n",
      "\t\t  - position: 329\n",
      "\t\t    ccd:phs\n",
      "\t\t  - position: 331\n",
      "\t\t    ccd:phs\n",
      "  - ligand: \n",
      "\tid: C\n",
      "\tccd: nag \n",
      "  - ligand: \n",
      "\tid: D\n",
      "\tccd: nag \n",
      "  - bond: \n",
      "\tatom1: B, 4, N\n",
      "\tatom2: C, 1, C\n",
      "  - bond: \n",
      "\tatom1: B, 16, N\n",
      "\tatom2: D, 1, C\n",
      "\n",
      "Boltz-1 input file has been created: /Users/ntw/Desktop/boltz1_inputs/P21731_Tirzepatide.yaml\n"
     ]
    }
   ],
   "source": [
    "##Modified from a ChatGPT draft.\n",
    "##Prompt: Using the uniprot ID as an input, make an input file for locally-installed \n",
    "## Boltz-1 with the protein sequence and post-translational modification data from uniprot, \n",
    "## including phosphorylation, methylation, glycosylation, and cross-linking. \n",
    "\n",
    "\n",
    "import requests\n",
    "\n",
    "# Dictionary to map common PTMs to their CCD identifiers\n",
    "ptm_to_ccd = {\n",
    "    \"phosphorylation\": {\n",
    "        \"phosphoserine\": \"phs\",     # Phosphorylation on serine\n",
    "        \"phosphothreonine\": \"pht\",  # Phosphorylation on threonine\n",
    "        \"phosphotyrosine\": \"phty\"   # Phosphorylation on tyrosine\n",
    "    },\n",
    "    \"methylation\": {\n",
    "        \"methylation (Lysine)\": \"me\",       # Methylation on lysine\n",
    "        \"methylation (Arginine)\": \"mearg\",  # Methylation on arginine\n",
    "        \"dimethylation (Lysine)\": \"me2\",    # Dimethylation on lysine\n",
    "        \"trimethylation (Lysine)\": \"me3\"    # Trimethylation on lysine\n",
    "    },\n",
    "    \"glycosylation\": {\n",
    "        \"N-linked Glycosylation\": \"ngly\",   # N-linked glycosylation\n",
    "        \"O-linked Glycosylation\": \"ogly\"    # O-linked glycosylation\n",
    "    },\n",
    "    \"acetylation\": {\n",
    "        \"acetylation (Lysine)\": \"ace\"  # Acetylation on lysine\n",
    "    },\n",
    "    \"ubiquitination\": {\n",
    "        \"ubiquitination\": \"ubiq\"  # Ubiquitin attachment\n",
    "    },\n",
    "    \"sumoylation\": {\n",
    "        \"sumoylation\": \"sumo\"  # SUMO attachment\n",
    "    },\n",
    "    \"crosslinking\": {\n",
    "        \"disulfide Bond\": \"dsb\",  # Disulfide bond\n",
    "        \"isopeptide Bond\": \"iso\"  # Isopeptide bond\n",
    "    }\n",
    "}\n",
    "\n",
    "seqGLP1R = { \"Tirzepatide\" : \"YAEGTFTSDYSIALDKIAQKAFVQWLIAGGPSSGAPPPS\", \\\n",
    "           \"Exenatide\" : \"HGEGTFTSDLSKQMEEEAVRLFIEWLKNGGPSSGAPPPS\", \\\n",
    "           \"Liraglutide\" : \"HAEGTFTSDVSSYLEGQAAKEEFIAWLVRGRG\", \\\n",
    "           \"Lixisenatide\" : \"HGEGTFTSDLSKQMEEEAVRLFIEWLKNGGPSSGAPPSKKKKKK\", \\\n",
    "           \"Albiglutide\" : \"HGEGTFTSDVSSYLEGQAAKEFIAWLVKGRHGEGTFTSDVSSYLEGQAAKEFIAWLVKGR\" \\\n",
    "           \"DAHKSEVAHRFKDLGEENFKALVLIAFAQYLQQCPFEDHVKLVNEVTEFAKTCVADESAE\" \\\n",
    "           \"NCDKSLHTLFGDKLCTVATLRETYGEMADCCAKQEPERNECFLQHKDDNPNLPRLVRPEV\" \\\n",
    "           \"DVMCTAFHDNEETFLKKYLYEIARRHPYFYAPELLFFAKRYKAAFTECCQAADKAACLLP\" \\\n",
    "           \"KLDELRDEGKASSAKQRLKCASLQKFGERAFKAWAVARLSQRFPKAEFAEVSKLVTDLTK\" \\\n",
    "           \"VHTECCHGDLLECADDRADLAKYICENQDSISSKLKECCEKPLLEKSHCIAEVENDEMPA\" \\\n",
    "           \"DLPSLAADFVESKDVCKNYAEAKDVFLGMFLYEYARRHPDYSVVLLLRLAKTYETTLEKC\" \\\n",
    "           \"CAAADPHECYAKVFDEFKPLVEEPQNLIKQNCELFEQLGEYKFQNALLVRYTKKVPQVST\" \\\n",
    "           \"PTLVEVSRNLGKVGSKCCKHPEAKRMPCAEDYLSVVLNQLCVLHEKTPVSDRVTKCCTES\" \\\n",
    "           \"LVNRRPCFSALEVDETYVPKEFNAETFTFHADICTLSEKERQIKKQTALVELVKHKPKAT\" \\\n",
    "           \"KEQLKAVMDDFAAFVEKCCKADDKETCFAEEGKKLVAASQAALGL\", \\\n",
    "            \"Dulaglutide\" : \"HGEGTFTSDVSSYLEEQAAKEFIAWLVKGGGGGGGSGGGGSGGGGSAESKYGPPCPPCPA\" \\\n",
    "           \"PEAAGGPSVFLFPPKPKDTLMISRTPEVTCVVVDVSQEDPEVQFNWYVDGVEVHNAKTKP\" \\\n",
    "           \"REEQFNSTYRVVSVLTVLHQDWLNGKEYKCKVSNKGLPSSIEKTISKAKGQPREPQVYTL\" \\\n",
    "           \"PPSQEEMTKNQVSLTCLVKGFYPSDIAVEWESNGQPENNYKTTPPVLDSDGSFFLYSRLT\" \\\n",
    "           \"VDKSRWQEGNVFSCSVMHEALHNHYTQKSLSLSLG\", \\\n",
    "           \"Semaglutide\" : \"HAEGTFTSDVSSYLEGQAAKEFIAWLVRGRG\"\n",
    "}\n",
    "\n",
    "#Dictionary to convert the number of entities to their albhabetic equivalents\n",
    "entity_to_id={i: chr(64 + i) for i in range(1, 27)}\n",
    "\n",
    "# Example usage: Retrieve CCD identifier for Phosphoserine\n",
    "ptm_type = \"phosphorylation\"\n",
    "modification = \"phosphoserine\"\n",
    "ccd_id = ptm_to_ccd[ptm_type][modification]\n",
    "#print(f\"The CCD identifier for {modification} is: {ccd_id}\")\n",
    "\n",
    "\n",
    "\n",
    "# Function to fetch protein data from UniProt\n",
    "def fetch_uniprot_data(uniprot_id):\n",
    "    # Define the UniProt API endpoint for retrieving protein data in JSON format\n",
    "    url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.json\"\n",
    "    response = requests.get(url)\n",
    "    \n",
    "    if response.status_code != 200:\n",
    "        print(f\"Error: Unable to fetch data for UniProt ID {uniprot_id}.\")\n",
    "        return None\n",
    "    \n",
    "    # Parse the JSON response\n",
    "    data = response.json()\n",
    "    \n",
    "    # Extract protein sequence and PTMs from the response\n",
    "    sequence = data['sequence']['value']\n",
    "    features = data.get('features', [])\n",
    "\n",
    "    #print(features)\n",
    "    \n",
    "    ptms = {\n",
    "        'phosphorylation': [],\n",
    "        'methylation': [],\n",
    "        'glycosylation': [],\n",
    "        'constraints': [],\n",
    "        'crosslinking': []\n",
    "    }\n",
    "    \n",
    "    # Collect the PTM data\n",
    "    for feature in features:\n",
    "        feature_type = feature['type']\n",
    "        description = feature.get('description', '').lower()\n",
    "        if feature['type'] == 'Modified residue':\n",
    "            if 'phosphothreonine' in description or 'phosphoserine' in description or 'phosphotyrosine' in description:\n",
    "                ptms['phosphorylation'].append({\n",
    "                'position': feature['location']['start']['value'],\n",
    "                'modification': description\n",
    "            })\n",
    "        elif feature['type'] == 'Methylation':\n",
    "            ptms['methylation'].append({\n",
    "                'position': feature['location']['start'],\n",
    "                'modification': feature['type']\n",
    "            })\n",
    "        elif feature['type'] == 'Glycosylation':\n",
    "            ptms['glycosylation'].append({\n",
    "                'position': feature['location']['start'],\n",
    "                'modification': feature['type'],\n",
    "                'description': feature['description']\n",
    "            })\n",
    "            ptms['constraints'].append({\n",
    "                'position': feature['location']['start'],\n",
    "                'modification': feature['type']\n",
    "            })\n",
    "            \n",
    "        elif feature['type'] == 'Cross-link':\n",
    "            ptms['crosslinking'].append({\n",
    "                'position': feature['location']['start'],\n",
    "                'modification': feature['type']\n",
    "            })\n",
    "    return sequence, ptms\n",
    "\n",
    "# Function to create the Boltz-1 input file\n",
    "def create_boltz1_input(entities, glp1r, sequence, ptms, output_filename):\n",
    "    gly_constraints=[]\n",
    "    # Create a string for the input file\n",
    "    input_content = \"version: 1\\n\"\n",
    "    input_content += \"sequences:\\n\"\n",
    "    input_content += \"  - protein:\\n\"\n",
    "    input_content += \"\\tid: \" + str(entity_to_id[entities]) + \"\\n\"\n",
    "    input_content += f\"\\tsequence: {seqGLP1R[glp1r]}\\n\"\n",
    "    input_content += \"\\tmsa: --use_msa_server\\n\"\n",
    "    entities+=1\n",
    "\n",
    "\n",
    "    \n",
    "    input_content += \"  - protein:\\n\"\n",
    "    input_content += \"\\tid: \" + str(entity_to_id[entities]) + \"\\n\"\n",
    "    input_content += f\"\\tsequence: {sequence}\\n\"\n",
    "    input_content += \"\\tmsa: --use_msa_server\\n\"\n",
    "    entity_for_mod=entities\n",
    "    # Add PTM information to the input\n",
    "\n",
    "    # Phosphorylation\n",
    "    if ptms['phosphorylation']:\n",
    "        input_content += \"\\tmodifications:\\n\"\n",
    "        for mod in ptms['phosphorylation']:\n",
    "            input_content += f\"\\t\\t  - position: {mod['position']}\"\n",
    "            input_content += \"\\n\"\n",
    "            input_content += \"\\t\\t    ccd:\"\n",
    "            input_content += ptm_to_ccd['phosphorylation'][mod['modification']]\n",
    "            input_content += \"\\n\"\n",
    "    \n",
    "    # Methylation\n",
    "    #if ptms['methylation']:\n",
    "    #    input_content += \"Methylation:\\n\"\n",
    "    #    for mod in ptms['methylation']:\n",
    "    #        input_content += f\"  Position: {mod['position']}, Modification: {mod['modification']}\\n\"\n",
    "    \n",
    "    # Glycosylation\n",
    "    if ptms['glycosylation']:\n",
    "        for mod in ptms['glycosylation']:\n",
    "            entities += 1\n",
    "            input_content += f\"  - ligand: \\n\" \n",
    "            #{mod['position']}\n",
    "            #Note: This assumes that glycosylation occurs between nitrogen and carbon,\n",
    "            # nitrogen from the amino acid residue and the glycan's anomeric carbon.\n",
    "            input_content += \"\\tid: \"  + str(entity_to_id[entities]) + \"\\n\"\n",
    "            if \"N-linked (GlcNAc\" in mod['description']:\n",
    "                input_content += \"\\tccd: \"\n",
    "                input_content += \"nag \\n\"\n",
    "                gly_constraints.append([str(entity_to_id[entity_for_mod]), mod['position']['value'], 'N'])\n",
    "                gly_constraints.append([str(entity_to_id[entities]), 1, 'C'])\n",
    "            if \"O-linked (GlcNAc\" in mod['description']:\n",
    "                input_content += \"\\tccd: \"\n",
    "                input_content += \"nag \\n\"\n",
    "                gly_constraints.append([str(entity_to_id[entitiy_for_mod]), mod['position']['value'], 'O'])\n",
    "                gly_constraints.append([str(entity_to_id[entities]), 1, 'C'])\n",
    "    \n",
    "    # Cross-linking\n",
    "    #if ptms['crosslinking']:\n",
    "    #    input_content += \"Cross-linking:\\n\"\n",
    "    #    for mod in ptms['crosslinking']:\n",
    "    #        input_content += f\"  Position: {mod['position']}, Modification: {mod['modification']}\\n\"\n",
    "    \n",
    "    if ptms['glycosylation'] or ptms['crosslinking']:\n",
    "        i=0\n",
    "        for mod in ptms['glycosylation']:\n",
    "            input_content += f\"  - bond: \\n\" \n",
    "            input_content += \"\\tatom1\" + \": \" + \\\n",
    "            str(gly_constraints[i]).replace('\\'', '').replace('[', '').replace(']', '') + \"\\n\"\n",
    "            i += 1\n",
    "            input_content += \"\\tatom2\" + \": \" + \\\n",
    "            str(gly_constraints[i]).replace('\\'', '').replace('[', '').replace(']', '') + \"\\n\"\n",
    "            i += 1\n",
    "\n",
    "            \n",
    "            \n",
    "    \n",
    "    # Write to the output file\n",
    "    print(input_content)\n",
    "    #print(gly_constraints)\n",
    "    with open(output_filename, 'w') as f:\n",
    "        f.write(input_content)\n",
    "    \n",
    "    print(f\"Boltz-1 input file has been created: {output_filename}\")\n",
    "\n",
    "# Example usage\n",
    "#   1) Assign the GLP1R agonist to pull its sequence from the seqGLP1R dictionary.\n",
    "# Options include: 'Tirzepatide', 'Exenatide', 'Liraglutide', \n",
    "#'Lixisenatide', 'Albiglutide', 'Dulaglutide', 'Semaglutide.' \n",
    "# Use seqGLP1R.keys() to see all options.\n",
    "\n",
    "glp1r=\"Tirzepatide\"\n",
    "\n",
    "#   2) Add the UniProt ID for the protein of interest. \n",
    "uniprot_id = 'P21731'    \n",
    "#   3) update the prefix for the directory to store the inputs.\n",
    "output_prefix = '/Users/ntw/Desktop/boltz1_inputs/'\n",
    "suffix= '.yaml'\n",
    "output_filename = output_prefix + uniprot_id + \"_\" + glp1r + suffix\n",
    "print(output_filename)\n",
    "\n",
    "# Fetch the UniProt data\n",
    "entities=0\n",
    "data = fetch_uniprot_data(uniprot_id)\n",
    "\n",
    "if data:\n",
    "    sequence, ptms = data\n",
    "    entities += 1\n",
    "\n",
    "    #print(ptms)\n",
    "    # Create the Boltz-1 input file\n",
    "    create_boltz1_input(entities, glp1r, sequence, ptms, output_filename)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fffe85cf-4c03-47f8-aa49-e0e602ead0ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cdc9cf6-2a4c-4547-904c-2660ddc7eac2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b24cf31-3cc1-43b6-949f-d323f8624d6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55f1bdc9-1846-4198-98d5-43c9bd09d667",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c4ca92-91fc-4b26-b08b-bb6ebdd4ce2c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
