{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5f75c74f-bed3-417a-bd40-04a04b3b06c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching data for UniProt ID: P11229\n",
      "Extracting protein information...\n",
      "[{'type': 'Chain', 'location': {'start': {'value': 1, 'modifier': 'EXACT'}, 'end': {'value': 460, 'modifier': 'EXACT'}}, 'description': 'Muscarinic acetylcholine receptor M1', 'featureId': 'PRO_0000069015'}, {'type': 'Topological domain', 'location': {'start': {'value': 1, 'modifier': 'EXACT'}, 'end': {'value': 22, 'modifier': 'EXACT'}}, 'description': 'Extracellular', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Transmembrane', 'location': {'start': {'value': 23, 'modifier': 'EXACT'}, 'end': {'value': 48, 'modifier': 'EXACT'}}, 'description': 'Helical; Name=1', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Topological domain', 'location': {'start': {'value': 49, 'modifier': 'EXACT'}, 'end': {'value': 62, 'modifier': 'EXACT'}}, 'description': 'Cytoplasmic', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Transmembrane', 'location': {'start': {'value': 63, 'modifier': 'EXACT'}, 'end': {'value': 84, 'modifier': 'EXACT'}}, 'description': 'Helical; Name=2', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Topological domain', 'location': {'start': {'value': 85, 'modifier': 'EXACT'}, 'end': {'value': 95, 'modifier': 'EXACT'}}, 'description': 'Extracellular', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Transmembrane', 'location': {'start': {'value': 96, 'modifier': 'EXACT'}, 'end': {'value': 121, 'modifier': 'EXACT'}}, 'description': 'Helical; Name=3', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Topological domain', 'location': {'start': {'value': 122, 'modifier': 'EXACT'}, 'end': {'value': 142, 'modifier': 'EXACT'}}, 'description': 'Cytoplasmic', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Transmembrane', 'location': {'start': {'value': 143, 'modifier': 'EXACT'}, 'end': {'value': 164, 'modifier': 'EXACT'}}, 'description': 'Helical; Name=4', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Topological domain', 'location': {'start': {'value': 165, 'modifier': 'EXACT'}, 'end': {'value': 185, 'modifier': 'EXACT'}}, 'description': 'Extracellular', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Transmembrane', 'location': {'start': {'value': 186, 'modifier': 'EXACT'}, 'end': {'value': 209, 'modifier': 'EXACT'}}, 'description': 'Helical; Name=5', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Topological domain', 'location': {'start': {'value': 210, 'modifier': 'EXACT'}, 'end': {'value': 366, 'modifier': 'EXACT'}}, 'description': 'Cytoplasmic', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Transmembrane', 'location': {'start': {'value': 367, 'modifier': 'EXACT'}, 'end': {'value': 390, 'modifier': 'EXACT'}}, 'description': 'Helical; Name=6', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Topological domain', 'location': {'start': {'value': 391, 'modifier': 'EXACT'}, 'end': {'value': 397, 'modifier': 'EXACT'}}, 'description': 'Extracellular', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Transmembrane', 'location': {'start': {'value': 398, 'modifier': 'EXACT'}, 'end': {'value': 420, 'modifier': 'EXACT'}}, 'description': 'Helical; Name=7', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Topological domain', 'location': {'start': {'value': 421, 'modifier': 'EXACT'}, 'end': {'value': 460, 'modifier': 'EXACT'}}, 'description': 'Cytoplasmic', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}, {'evidenceCode': 'ECO:0007744', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Region', 'location': {'start': {'value': 225, 'modifier': 'EXACT'}, 'end': {'value': 256, 'modifier': 'EXACT'}}, 'description': 'Disordered', 'evidences': [{'evidenceCode': 'ECO:0000256', 'source': 'SAM', 'id': 'MobiDB-lite'}]}, {'type': 'Region', 'location': {'start': {'value': 274, 'modifier': 'EXACT'}, 'end': {'value': 297, 'modifier': 'EXACT'}}, 'description': 'Disordered', 'evidences': [{'evidenceCode': 'ECO:0000256', 'source': 'SAM', 'id': 'MobiDB-lite'}]}, {'type': 'Region', 'location': {'start': {'value': 310, 'modifier': 'EXACT'}, 'end': {'value': 351, 'modifier': 'EXACT'}}, 'description': 'Disordered', 'evidences': [{'evidenceCode': 'ECO:0000256', 'source': 'SAM', 'id': 'MobiDB-lite'}]}, {'type': 'Compositional bias', 'location': {'start': {'value': 231, 'modifier': 'EXACT'}, 'end': {'value': 248, 'modifier': 'EXACT'}}, 'description': 'Polar residues', 'evidences': [{'evidenceCode': 'ECO:0000256', 'source': 'SAM', 'id': 'MobiDB-lite'}]}, {'type': 'Site', 'location': {'start': {'value': 170, 'modifier': 'EXACT'}, 'end': {'value': 170, 'modifier': 'EXACT'}}, 'description': 'Subtype-specific residue that binds to snake venom muscarinic toxin 7', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}]}, {'type': 'Site', 'location': {'start': {'value': 172, 'modifier': 'EXACT'}, 'end': {'value': 172, 'modifier': 'EXACT'}}, 'description': 'Binds to snake venom muscarinic toxin 7', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}]}, {'type': 'Site', 'location': {'start': {'value': 174, 'modifier': 'EXACT'}, 'end': {'value': 174, 'modifier': 'EXACT'}}, 'description': 'Subtype-specific residue that binds to snake venom muscarinic toxin 7', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}]}, {'type': 'Site', 'location': {'start': {'value': 397, 'modifier': 'EXACT'}, 'end': {'value': 397, 'modifier': 'EXACT'}}, 'description': 'Subtype-specific residue that binds to snake venom muscarinic toxin 7', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}]}, {'type': 'Site', 'location': {'start': {'value': 401, 'modifier': 'EXACT'}, 'end': {'value': 401, 'modifier': 'EXACT'}}, 'description': 'Subtype-specific residue that binds to snake venom muscarinic toxin 7', 'evidences': [{'evidenceCode': 'ECO:0000269', 'source': 'PubMed', 'id': '32646996'}]}, {'type': 'Modified residue', 'location': {'start': {'value': 230, 'modifier': 'EXACT'}, 'end': {'value': 230, 'modifier': 'EXACT'}}, 'description': 'Phosphothreonine', 'evidences': [{'evidenceCode': 'ECO:0000250', 'source': 'UniProtKB', 'id': 'P12657'}]}, {'type': 'Modified residue', 'location': {'start': {'value': 428, 'modifier': 'EXACT'}, 'end': {'value': 428, 'modifier': 'EXACT'}}, 'description': 'Phosphothreonine', 'evidences': [{'evidenceCode': 'ECO:0000255'}]}, {'type': 'Modified residue', 'location': {'start': {'value': 451, 'modifier': 'EXACT'}, 'end': {'value': 451, 'modifier': 'EXACT'}}, 'description': 'Phosphoserine', 'evidences': [{'evidenceCode': 'ECO:0000255'}]}, {'type': 'Modified residue', 'location': {'start': {'value': 455, 'modifier': 'EXACT'}, 'end': {'value': 455, 'modifier': 'EXACT'}}, 'description': 'Phosphothreonine', 'evidences': [{'evidenceCode': 'ECO:0000255'}]}, {'type': 'Modified residue', 'location': {'start': {'value': 457, 'modifier': 'EXACT'}, 'end': {'value': 457, 'modifier': 'EXACT'}}, 'description': 'Phosphoserine', 'evidences': [{'evidenceCode': 'ECO:0000255'}]}, {'type': 'Glycosylation', 'location': {'start': {'value': 2, 'modifier': 'EXACT'}, 'end': {'value': 2, 'modifier': 'EXACT'}}, 'description': 'N-linked (GlcNAc...) asparagine', 'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU00498'}], 'featureId': ''}, {'type': 'Glycosylation', 'location': {'start': {'value': 12, 'modifier': 'EXACT'}, 'end': {'value': 12, 'modifier': 'EXACT'}}, 'description': 'N-linked (GlcNAc...) asparagine', 'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU00498'}], 'featureId': ''}, {'type': 'Disulfide bond', 'location': {'start': {'value': 98, 'modifier': 'EXACT'}, 'end': {'value': 178, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU00521'}]}, {'type': 'Alternative sequence', 'location': {'start': {'value': 455, 'modifier': 'EXACT'}, 'end': {'value': 460, 'modifier': 'EXACT'}}, 'description': 'in isoform 2', 'evidences': [{'evidenceCode': 'ECO:0000303', 'source': 'Reference', 'id': 'Ref.4'}], 'featureId': 'VSP_056651', 'alternativeSequence': {}}, {'type': 'Sequence conflict', 'location': {'start': {'value': 173, 'modifier': 'EXACT'}, 'end': {'value': 173, 'modifier': 'EXACT'}}, 'description': 'in Ref. 3; CAA33334', 'evidences': [{'evidenceCode': 'ECO:0000305'}], 'alternativeSequence': {'originalSequence': 'V', 'alternativeSequences': ['M']}}, {'type': 'Helix', 'location': {'start': {'value': 27, 'modifier': 'EXACT'}, 'end': {'value': 52, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 54, 'modifier': 'EXACT'}, 'end': {'value': 56, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 59, 'modifier': 'EXACT'}, 'end': {'value': 76, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 78, 'modifier': 'EXACT'}, 'end': {'value': 88, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 95, 'modifier': 'EXACT'}, 'end': {'value': 128, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Turn', 'location': {'start': {'value': 130, 'modifier': 'EXACT'}, 'end': {'value': 132, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 133, 'modifier': 'EXACT'}, 'end': {'value': 135, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 139, 'modifier': 'EXACT'}, 'end': {'value': 168, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 182, 'modifier': 'EXACT'}, 'end': {'value': 184, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZG4'}]}, {'type': 'Helix', 'location': {'start': {'value': 186, 'modifier': 'EXACT'}, 'end': {'value': 196, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 198, 'modifier': 'EXACT'}, 'end': {'value': 219, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 362, 'modifier': 'EXACT'}, 'end': {'value': 390, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Beta strand', 'location': {'start': {'value': 391, 'modifier': 'EXACT'}, 'end': {'value': 393, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6WJC'}]}, {'type': 'Helix', 'location': {'start': {'value': 397, 'modifier': 'EXACT'}, 'end': {'value': 408, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 410, 'modifier': 'EXACT'}, 'end': {'value': 421, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 423, 'modifier': 'EXACT'}, 'end': {'value': 437, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6ZFZ'}]}, {'type': 'Helix', 'location': {'start': {'value': 439, 'modifier': 'EXACT'}, 'end': {'value': 441, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '6OIJ'}]}]\n",
      "Chain\n",
      "Topological domain\n",
      "Transmembrane\n",
      "Topological domain\n",
      "Transmembrane\n",
      "Topological domain\n",
      "Transmembrane\n",
      "Topological domain\n",
      "Transmembrane\n",
      "Topological domain\n",
      "Transmembrane\n",
      "Topological domain\n",
      "Transmembrane\n",
      "Topological domain\n",
      "Transmembrane\n",
      "Topological domain\n",
      "Region\n",
      "Region\n",
      "Region\n",
      "Compositional bias\n",
      "Site\n",
      "Site\n",
      "Site\n",
      "Site\n",
      "Site\n",
      "Modified residue\n",
      "phosphothreonine\n",
      "Phosphothreonine\n",
      "Modified residue\n",
      "phosphothreonine\n",
      "Phosphothreonine\n",
      "Modified residue\n",
      "phosphoserine\n",
      "Phosphoserine\n",
      "Modified residue\n",
      "phosphothreonine\n",
      "Phosphothreonine\n",
      "Modified residue\n",
      "phosphoserine\n",
      "Phosphoserine\n",
      "Glycosylation\n",
      "Glycosylation\n",
      "Disulfide bond\n",
      "Alternative sequence\n",
      "Sequence conflict\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Turn\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Beta strand\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Helix\n",
      "Generating AlphaFold input JSON...\n",
      "Saving input to alphafold_input_P11229.json...\n",
      "AlphaFold input JSON for P11229 saved as alphafold_input_P11229.json\n"
     ]
    }
   ],
   "source": [
    "## ChatGPT: Provide a script to automatically generate AlphaFold 3 input json \n",
    "## files for a protein given the uniprot identifier for that protein, pulling \n",
    "## and including post-translational modification data from the Uniprot entry, \n",
    "## including phosphorylation, glycosylation, acetylation, and methylation. \n",
    "## The glycosylation format should be compatible with a local installation of AlphaFold 3\n",
    "\n",
    "import requests\n",
    "import json\n",
    "\n",
    "# Function to fetch data from UniProt\n",
    "def fetch_uniprot_data(uniprot_id):\n",
    "    # Define the UniProt API endpoint\n",
    "    url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.json\"\n",
    "\n",
    "    # Send the request to the UniProt API\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # Check if the response is valid\n",
    "    if response.status_code != 200:\n",
    "        raise Exception(f\"Failed to retrieve data for UniProt ID {uniprot_id}. HTTP Status Code: {response.status_code}\")\n",
    "\n",
    "    # Return the JSON response\n",
    "    return response.json()\n",
    "\n",
    "# Function to extract relevant data from the UniProt entry\n",
    "def extract_protein_info(uniprot_data):\n",
    "    #print(uniprot_data)\n",
    "    # Extract protein sequence\n",
    "    #print(uniprot_data['sequence']['value'])\n",
    "    sequence = uniprot_data['sequence']['value']\n",
    "\n",
    "    # Extract PTM data (phosphorylation, glycosylation, acetylation, and methylation)\n",
    "    ptms = []\n",
    "    print(uniprot_data.get('features', []))\n",
    "    for feature in uniprot_data.get('features', []):\n",
    "        feature_type = feature['type']\n",
    "        description = feature.get('description', '').lower()\n",
    "        print(feature_type)\n",
    "\n",
    "        # Collect PTM features of interest\n",
    "        if feature_type == 'Modified residue':\n",
    "            # Check for specific PTMs\n",
    "            print(description)\n",
    "            if 'phosphothreonine' in description or 'phosphoserine' in description or 'phosphotyrosine' in description:\n",
    "                ptms.append({\n",
    "                    'type': 'Phosphorylation',\n",
    "                    'description': feature.get('description', 'No description'),\n",
    "                    'location': feature['location']['start']\n",
    "                })\n",
    "                print(feature.get('description', 'No description'))\n",
    "            elif 'acetylation' in description:\n",
    "                ptms.append({\n",
    "                    'type': 'Acetylation',\n",
    "                    'description': feature.get('description', 'No description'),\n",
    "                    'location': feature['location']\n",
    "                })\n",
    "            elif 'methylation' in description:\n",
    "                ptms.append({\n",
    "                    'type': 'Methylation',\n",
    "                    'description': feature.get('description', 'No description'),\n",
    "                    'location': feature['location']\n",
    "                })\n",
    "\n",
    "        # If it's a cross-link, include it as a general PTM type\n",
    "        elif feature_type == 'Glycosylation':\n",
    "            ptms.append({\n",
    "                'type': 'Glycosylation',\n",
    "                'description': feature.get('description', 'No description'),\n",
    "                'location': feature['location']['start'],\n",
    "                'glycan_type': 'N-linked'  # Default assumption, could be updated to O-linked if specified\n",
    "                })\n",
    "        \n",
    "        elif feature_type == 'Disulfide bond':\n",
    "            ptms.append({\n",
    "                'type': 'Disulfide bond',\n",
    "                'description': feature.get('description', 'No description'),\n",
    "                'location': feature['location']['start'],\n",
    "                'end': feature['location']['end']\n",
    "            })\n",
    "\n",
    "    # Extract other annotations or metadata (e.g., protein name)\n",
    "    protein_name = uniprot_data.get('proteinDescription', {}).get('recommendedName', {}).get('fullName', {}).get('value', 'Unknown')\n",
    "\n",
    "    return {\n",
    "        'sequence': sequence,\n",
    "        'ptms': ptms,\n",
    "        'protein_name': protein_name\n",
    "    }\n",
    "\n",
    "# Function to generate the AlphaFold input JSON structure\n",
    "def generate_alphafold_input_json(protein_info):\n",
    "    # Construct the AlphaFold input JSON structure\n",
    "    alphafold_input = {\n",
    "        \"sequence\": protein_info['sequence'],\n",
    "        \"name\": protein_info['protein_name'],\n",
    "        \"ptms\": protein_info['ptms'],\n",
    "        \"metadata\": {\n",
    "            \"description\": \"Protein sequence and PTM data for AlphaFold\",\n",
    "            \"source\": \"UniProt\"\n",
    "        }\n",
    "    }\n",
    "\n",
    "    # Format glycosylation PTMs to match the AlphaFold 3 format\n",
    "    for ptm in alphafold_input['ptms']:\n",
    "        if ptm['type'] == 'Glycosylation':\n",
    "            # Check if glycosylation is N-linked or O-linked based on the description\n",
    "            if 'N-linked' in ptm['description']:\n",
    "                ptm['glycan_type'] = 'N-linked'\n",
    "            elif 'O-linked' in ptm['description']:\n",
    "                ptm['glycan_type'] = 'O-linked'\n",
    "            else:\n",
    "                ptm['glycan_type'] = 'Unknown'  # You could refine this logic further\n",
    "\n",
    "    return alphafold_input\n",
    "\n",
    "# Function to save the JSON to a file\n",
    "def save_to_json_file(data, filename):\n",
    "    with open(filename, 'w') as json_file:\n",
    "        json.dump(data, json_file, indent=4)\n",
    "\n",
    "# Main function to orchestrate the process\n",
    "def main(uniprot_id, output_filename):\n",
    "    # Step 1: Fetch UniProt data\n",
    "    print(f\"Fetching data for UniProt ID: {uniprot_id}\")\n",
    "    uniprot_data = fetch_uniprot_data(uniprot_id)\n",
    "\n",
    "    # Step 2: Extract relevant protein information\n",
    "    print(\"Extracting protein information...\")\n",
    "    protein_info = extract_protein_info(uniprot_data)\n",
    "\n",
    "    # Step 3: Generate AlphaFold input JSON\n",
    "    print(\"Generating AlphaFold input JSON...\")\n",
    "    alphafold_input = generate_alphafold_input_json(protein_info)\n",
    "\n",
    "    # Step 4: Save to JSON file\n",
    "    print(f\"Saving input to {output_filename}...\")\n",
    "    save_to_json_file(alphafold_input, output_filename)\n",
    "\n",
    "    print(f\"AlphaFold input JSON for {uniprot_id} saved as {output_filename}\")\n",
    "\n",
    "# Example usage\n",
    "if __name__ == \"__main__\":\n",
    "    # Specify UniProt ID and output filename\n",
    "    uniprot_id = \"P11229\"  # UniProt ID for SOD1\n",
    "    output_filename = \"alphafold_input_P11229.json\"\n",
    "\n",
    "    # Run the main function\n",
    "    main(uniprot_id, output_filename)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dead4f45-fb33-48de-bfab-8fca6801006f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "317f5d5c-4751-4d17-8d81-dd8716b7eb93",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
